### PR TITLE
Embedded elf guests for forks

### LINKED ISSUE #1476: New compilation and read ELF from files system
checkout esad/new-compilation-system branch

The goal we have with this issue is to read ELfs from files instead of by building and putting it as a constant in the binary. With this change, our provers will be able to prove with any ELF at any point in time. (If we don't do this forking becomes very time dependent for deployment). However, for test builds, we would like to use the newly compiled ELF, this is already implemented in the branch. So for non-testnet builds, we'd like to not see risc0 guests get compiled.

Ideally, we would like a new directory in project root, `elfs/`, which would have two subdirectories `risc0/` and `sp1`. Each will have subdirectories `testnet/` `mainnet/` and the ELFs will be inside these subdirectories. The naming convention can be 0.elf 1.elf ... for different forks. However, it's ok to implement them differently if implementor wishes.

Now, notice that for the production builds, we now don't depend on `citrea-risc0`, the crate that holds our ZK guests, and we can use the same binary on both testnet and mainnet. (This would become a problem, let's say, if we put genesis state root of the rollup as a constant in the light client proof guest. We would need to build two guests, hence two binaries. But we can do 2 ELFs and one binary, which we point at which elfs to load. see #1444 ) 

A new CLI arg --testnet is needed to indicate that the node is going to be run on Citrea testnet. (absence of --testnet indicates mainnet run.). This would load ELFs from `elfs/../testnet`.

Now that we get our provers to load multiple ELFs, we'd also need to change prover service trait so that the prover runner can signal which ELF will be used by proving. To actually decide on which ELF to use for proving see below:

- the batch prover will look at the L2 range it is about to prove, if at any point fork x is activated, it will use the elf for fork x.
- the light client prover will always use the latest ELF, we can change this behaviour later. I am not sure at this moment.

The full node behaviour of loading up method id's as constants are now also gone, the full nodes will load ELF's at the corresponding directory and calculate method id's on their own. They can discard the ELF's afterwards.

In the PR, please also go back to our latest release, do a reproducible build of the batch proof guest, and put it under testnet's first elf.

### PR DESCRIPTION
**Pull request by rakanalh**
# Description
`make build` now builds the ELF guest codes separately using `make build-risc0` and `make build-sp1`.

This basically removes the fact that we build the guest code along with the binary. Now, the guest code is loaded from `resources/guests` based on what we want to load.

That being said, each fork should add the ELF to the designated directory with `1.elf`, `2.elf` .. etc for each fork we end up introducing. The names can change of course but that's how it goes for now.

# Development workflow

Previously, we used to have the guest code already built within the binary. If you run:
- `make build` 
- `make test`

Both will act the same way by building the guest code and provide a binary that has the guest codes needed.

If you are running testing a specific build for the guest code, you can use:
- `make -C guests/risc0 batch-proof-bitcoin`
- `make -C guests/risc0 batch-proof-mock`
- `make -C guests/risc0 light-client-bitcoin`
- `make -C guests/risc0 light-client-mock`

# TODO
- [x] Prevent `build.rs` for risc0 from running.
- [x] Every time a fork activates, we should upload the newly activated guest code to Bonsai
- [x] Override `network` if `--testnet` is used.
- [x] Pass `Network` argument

## Linked Issues
- Fixes #1476

### CODE REVIEW

File: Makefile
Comment (rakanalh) [inline Makefile]: Seems like the build command has `--target-dir` to specify where the elfs should go... but i couldn't find it when running `cargo risczero build` locally.
https://github.com/risc0/risc0/blob/v1.1.3/risc0/cargo-risczero/src/commands/build.rs#L56

Seems this is unreleased yet?

File: bin/citrea/src/main.rs
Comment (rakanalh) [inline bin/citrea/src/main.rs:34]: TODO: Pass this around

File: Makefile
Comment (eyusufatik) [reply Makefile]: are you sure you're on 1.1.3?

Comment (rakanalh) [reply Makefile]:
```
cargo risczero --version
cargo-risczero 1.1.3
```

File: bin/citrea/src/utils.rs
Comment (eyusufatik) [inline bin/citrea/src/utils.rs]: didn't read so carefully but just a heads up

the batch prover will choose which elf to use with this algo:

1- get sequencer commitments last l2 height.
2- get spec at that height
3- use corresponding elf

the light client prover will need use this algo:

1- get last l2 height in batch proof output
2- get spec at that height
3- use the elf for that spec
4- set the batch proof method id input from the batch proof elf at that spec 

(this is a bit due to change, we might change how we handle the light client case, even inputs might change)

**Comment by eyusufatik** [comment]
> Every time a fork activates, we should upload the newly activated guest code to Bonsai

btw just pass down elf from runner > prover service > zkvm host

I think bonsai doesnt upload all the time since it checks if elf exists on remote

File: crates/risc0/src/host.rs
Comment (eyusufatik) [inline crates/risc0/src/host.rs]: image_id is irrelevant at this point if we are passing elfs on `run`

File: crates/light-client-prover/src/da_block_handler.rs
Comment (eyusufatik) [inline crates/light-client-prover/src/da_block_handler.rs]: can you put this log back with debug! ?

File: crates/batch-prover/src/proving.rs
Comment (eyusufatik) [inline crates/batch-prover/src/proving.rs:234]: the grouping that'll be applied to these sequencer commitments are important.

let's say there is 3 items i n`sequencer_commmitments` vec, and only the last one has the height which the next fork activates.

down in execution, if these guys are grouped like [(1,2), (3)] than we can prove the first group  elf n-1 and the last group with elf 3. 

or else there will be confusion when we need to pull batch proof method ids based on `output.last_l2_height` in light client circuit

**Review by eyusufatik — APPROVED** [review]
looks good. I'll need another round of reviewing since I've only reviewed since my last review this time. needs a small change though

**Review by eyusufatik — CHANGES_REQUESTED** [review]
I meant to request changes

File: crates/risc0/src/host.rs
Comment (rakanalh) [reply crates/risc0/src/host.rs]:
```
        receipt.verify(self.image_id)?;
```
In host, we use the image ID to verify receipt apparently.

Comment (eyusufatik) [reply crates/risc0/src/host.rs]: yes but image id can be calculated from the elf passed wherever needed.

File: bin/citrea/Cargo.toml
Comment (rakanalh) [inline bin/citrea/Cargo.toml]: Place correct revision after companion PR is merged.

File: crates/batch-prover/src/proving.rs
Comment (rakanalh) [reply crates/batch-prover/src/proving.rs:234]: I have implemented a change in `break_sequencer_commitments` to not only account for state diff threshold, but also forks. Commitments will be grouped by fork.

Comment (rakanalh) [reply crates/batch-prover/src/proving.rs:234]: One thing to note, is that with:
`GroupCommitment::SingleShot`: we don't account for forks
`GroupCommitment::OneByOne`: Each commitment is split by default so no need to account for forks
`break_sequencer_commitment`: Accounts for forks.

File: bin/citrea/src/rollup/bitcoin.rs
Comment (eyusufatik) [inline bin/citrea/src/rollup/bitcoin.rs]: (can be a new issue) 

since the zkvm adapters now don't need elfs on ::new()

we can have the same create_prover_service for light client prover and batch prover, where the da verifier is an input

File: crates/batch-prover/src/da_block_handler.rs
Comment (eyusufatik) [inline crates/batch-prover/src/da_block_handler.rs]: check l2_end_block_number here

File: crates/batch-prover/src/proving.rs
Comment (eyusufatik) [inline crates/batch-prover/src/proving.rs:236]: this is good, however, we'll need to generate V1 input if we prove for older elfs on testnet or mainnet

File: crates/light-client-prover/src/da_block_handler.rs
Comment (eyusufatik) [inline crates/light-client-prover/src/da_block_handler.rs]: we'll need to get method id by batch_proof_output.last_l2_height's spec id

File: crates/prover-services/src/parallel/mod.rs
Comment (eyusufatik) [inline crates/prover-services/src/parallel/mod.rs:116]: what if different proofs in the queue require different elfs

File: crates/batch-prover/src/da_block_handler.rs
Comment (eyusufatik) [reply crates/batch-prover/src/da_block_handler.rs]: I meant get with l2_end_block_number

File: crates/prover-services/src/parallel/mod.rs
Comment (rakanalh) [reply crates/prover-services/src/parallel/mod.rs:116]: The batch prover should break up commitments and therefore we'll have different inputs in which the inputs are related to non-forking commitments.

File: bin/citrea/src/rollup/bitcoin.rs
Comment (rakanalh) [reply bin/citrea/src/rollup/bitcoin.rs]: Done

File: crates/light-client-prover/src/da_block_handler.rs
Comment (rakanalh) [reply crates/light-client-prover/src/da_block_handler.rs]: Done

File: crates/batch-prover/src/proving.rs
Comment (rakanalh) [reply crates/batch-prover/src/proving.rs:236]: Will be done in: https://github.com/chainwayxyz/citrea/issues/1538

File: resources/guests/risc0/testnet/batch-0.elf
Comment (eyusufatik) [inline resources/guests/risc0/testnet/batch-0.elf:1]: please make sure image id is the same with our old image id

File: .github/workflows/checks.yml
Comment (eyusufatik) [inline .github/workflows/checks.yml]: might want to remove this

**Review by eyusufatik — APPROVED** [review]

File: resources/guests/risc0/testnet/batch-0.elf
Comment (rakanalh) [reply resources/guests/risc0/testnet/batch-0.elf:1]: Built the code on 0.5.4 release using:
```
REPR_GUEST_BUILD=1 make build
```

Calculated the image ID for bitcoin prover:
```
> cp target/riscv-guest/riscv32im-risc0-zkvm-elf/docker/citrea_bitcoin_prover/bitcoin_da resources/guests/risc0/testnet/batch-0.elf

> r0vm --elf resources/guests/risc0/testnet/batch-0.elf --id
3631d90630a3f0deb47f3a3411fe6e7ede1b0d86ad4216c75041e1a2020f009f
```
Checked against calculated value in `guests.rs`:

```rust
    println!(
        "{:?}",
        citrea::guests::BATCH_PROOF_TESTNET_GUESTS
            .get(&SpecId::Genesis)
            .unwrap()
            .0
    );
```

Result: 
```
Digest(3631d90630a3f0deb47f3a3411fe6e7ede1b0d86ad4216c75041e1a2020f009f)
```

File: Makefile
Comment (yaziciahmet) [inline Makefile:21]: I think we should skip building sp1 for now.

File: bin/citrea/src/main.rs
Comment (yaziciahmet) [inline bin/citrea/src/main.rs:40]: I don't see dev being used anywhere

File: bin/citrea/src/rollup/bitcoin.rs
Comment (yaziciahmet) [inline bin/citrea/src/rollup/bitcoin.rs:243]: nice!

**Review by yaziciahmet — COMMENTED** [review]
LGTM! not blocking the pr, just have 1 comment.

File: bin/citrea/src/main.rs
Comment (rakanalh) [reply bin/citrea/src/main.rs:40]: https://github.com/chainwayxyz/citrea/pull/1495/files/36bf252ef863ee0e0a07cf9251a83266b7ddcabc#diff-c1783a446ab178e59e9ad736b1364049f88a3efac218f37793c2ca081b5489ffR153

Comment (yaziciahmet) [reply bin/citrea/src/main.rs:40]: ok im blind, ty :D

**Review by yaziciahmet — APPROVED** [review]

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**
